pkg/cloudprovider/nodecontroller/podevictor.go

```diff
@@ -0,0 +1,99 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nodecontroller
+
+import (
+	"sync"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+
+	"github.com/golang/glog"
+)
+
+type empty struct{}
+
+// Entity responsible for evicting Pods from inserted Nodes. It uses RateLimiter to avoid
+// evicting everything at once.
+type PodEvictor struct {
+	lock         sync.Mutex
+	evictQueue   []string
+	evictedNodes util.StringSet
+
+	deletingPodsRateLimiter util.RateLimiter
+}
+
+// Creates new PodEvictor which will use given RateLimiter to oversee eviction.
+func NewPodEvictor(deletingPodsRateLimiter util.RateLimiter) *PodEvictor {
+	return &PodEvictor{
+		evictQueue:              make([]string, 0),
+		evictedNodes:            util.NewStringSet(),
```

Comment:
No. The main reason is to have a 'history' of evicted nodes, to avoid no-op node evictions. FIFO is enough to ensure fairness, but I believe it'd be bad if having one evicted node cause hitting RateLimiter limit.